ddtrace/profiling/collector/_memalloc_heap.c

```diff
@@ -351,7 +361,93 @@ memalloc_heap(void)
      * New allocations will go into the secondary freezer.allocs_m map and allocations
      * tracked in allocs_m which are freed will be added to a list to be removed when
      * the profiler is thawed. */
-    PyObject* heap_list = memalloc_heap_map_export(global_heap_tracker.allocs_m);
+    
+    /* Calculate total number of samples: live + freed */
+    size_t live_count = memalloc_heap_map_size(global_heap_tracker.allocs_m);
+    size_t freed_count = global_heap_tracker.allocation_list.count;
+    size_t total_count = live_count + freed_count;
+    
+    PyObject* heap_list = PyList_New(total_count);
+    if (heap_list == NULL) {
+        heap_tracker_thaw(&global_heap_tracker);
+        return NULL;
+    }
+
+    int list_index = 0;
+    
+    /* First, iterate over live samples using the new iterator API */
+    memalloc_heap_map_iter_t* it = memalloc_heap_map_iter_new(global_heap_tracker.allocs_m);
+    if (it) {
+        void* key;
+        traceback_t* tb;
+        
+        while (memalloc_heap_map_iter_next(it, &key, &tb)) {
+            if (list_index >= total_count) {
+                break;
+            }
+            
+            PyObject* tb_and_info = PyTuple_New(5);
+            if (tb_and_info == NULL) {
+                continue;
+            }
+            
+            PyTuple_SET_ITEM(tb_and_info, 0, traceback_to_tuple(tb));
+            PyTuple_SET_ITEM(tb_and_info, 1, PyLong_FromSize_t(tb->size));
+            PyTuple_SET_ITEM(tb_and_info, 2, Py_True); /* in_use = True for live samples */
+            Py_INCREF(Py_True);
+            PyTuple_SET_ITEM(tb_and_info, 3, PyLong_FromSize_t(tb->count));
+            PyTuple_SET_ITEM(tb_and_info, 4, tb->reported ? Py_True : Py_False);
+            Py_INCREF(tb->reported ? Py_True : Py_False);
```

Comment:
An idea I had while looking at this last week: Maybe we can have the in-use size and allocated size computed here rather than in `memalloc.py`? Right now `memalloc.py` has two implementation of that computation, one in `snapshot` and one in `test_snapshot`. But in theory we could do that here since we already have all the info we need.

So basically, instead of creating a tuple `(traceback, size, in_use, count, reported)`, create a tuple `(traceback, in_use_size, alloc_size, count)`. WDYT? Let me know if you want more detail.